### DISCUSSION TITLE
feat: Add HTTP Cache Middleware

### DISCUSSION THREAD

**Discussion by darkweak** [confetti-framework/confetti#143] [Ideas]
Hello there, first of all, thank you for this Laravel-like Golang  web app framework.

I already wrote an HTTP cache system ([called Souin](https://github.com/darkweak/souin)) used by the caddy cache-handler module and roadrunner, compatible with Træfik, Tyk and many other reverse-proxies/API Gateway. It supports the RFC-7234, can partially cache graphQL request, invalidate using xkeys/ykeys like Varnish. It also implements the Fastly purge using the Surrogate-key header. It can store and invalidate a CDN (cloudflare, fastly, Akamai), and it supports the Cache-Status RFC directive. It implements two in-memory/fs storages (badger & nutsdb) and two distributed storages (olric & etcd) that are fully configurable. The keys can be tweaked (e.g. serve the same cached css for multiple domains) and we can change the Cache-Status name through the configuration. And in the next release, all responses will be streamed and support the ESI tags.

What do you think about implementing it in the cache repository and make something like the cache-handler, the Souin repo could be the stable development repository and the `confetti-framework/confetti` could contains the stable production code.

Let me know your preference about that. ✌️

**Comment by reindert-vetter** [comment]
What a great package!

If it is passed directly into the framework, it must also be fully documented in the documentation. What does bother me is that the go.mod is very large. go.sum of the framework would be quadrupled. That while it is mainly only used in server site generated webpages (if I'm not mistaken).

That's why I think we'd better make it easy to use it when needed. Then there could be a short explanation with a link in the documentation.

**Reply by darkweak** [reply]
The only added deps in the `go.mod` would be `github.com/darkweak/souin` and maybe `zap` (but removed in the future to have a generic logger compatible with syslog, zap, ...) but I wonder if the go.sum will be really changed a lot. So it won't change a lot the codebase IMHO.

It could be something like `roadrunner` then with a third-party but official HTTP cache middleware.   https://roadrunner.dev/docs/middleware-cache/2.x/en so I could write the documentation about the installation and configuration with some example and put a link in the documentation that redirect to the Souin repository to go further. So I/we could just have to maintain the `souin/plugins/confetti`repository.

**Reply by reindert-vetter** [reply]
Maybe it's okay. I'd have to investigate that. I love the idea of the `souin/plugins/confettirepository` repository.
@bosix and @Wulfheart, would this be something you would use?

**Reply by Wulfheart** [reply]
I currently don't develop using confetti.
